### PR TITLE
DDP-4648 studymanager firewall

### PR DESCRIPTION
**Pull request by zyme**
## Context

DDP-4648 is about allowing traffic between study-server and study-manager while maintaining an external firewall for study-manager.  Instance tagging is not supported by app engine standard, although it is supported by flex.

The current POC uses [serverless VPC access](https://cloud.google.com/appengine/docs/standard/python3/connecting-vpc) and internal static IP in study-manager to allow traffic from study-server.

_I have not filled in the rest of the checklist yet_

## Checklist

- [ ] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/config/application.conf.ctmpl [hunk 1]
```diff
@@ -93,7 +93,7 @@
     {{if eq $environment "dev"}}
         "baseTestUrl": "https://pepper-dev.datadonationplatform.org",
         "baseWebDriverUrl": "https://basil-dev.datadonationplatform.org",
-        "dsmBaseUrl": "https://dsm-dev.datadonationplatform.org:443"
+        "dsmBaseUrl": "https://10.128.0.6:443"
```
Comment (zyme) [inline pepper-apis/config/application.conf.ctmpl]: Now-static internal IP for dev DSM.  Would apply this pattern across all envs.

File: pepper-apis/src/appengine/DataDonationPlatform.yaml [hunk 2]
```diff
@@ -13,8 +13,10 @@ automatic_scaling:
   max_concurrent_requests: 10 #default is 10 . Max is 80. GCP recommends 10 max value.
   max_pending_latency: 60ms #Default is 30ms. Max time a request is allowed to wait in queue.
   min_pending_latency: 60ms #Default is 30ms. Threshold used to scale down
-network:
-  instance_tag: study-server # necessary for firewall rules so that DSM will accept https requests from pepper
+
+vpc_access_connector:
+  # temporary proof-of-concept.  how to parameterize $PROJECT_ID without copy-pasting yaml files?
+  name: "projects/broad-ddp-dev/locations/us-central1/connectors/appengine-connector"
```
Comment (zyme) [inline pepper-apis/src/appengine/DataDonationPlatform.yaml]: Suggestions for how to parameterize `$PROJECT_ID`?  Git templating?

**Review by SimoneMaiwald — DISMISSED** [review]

File: pepper-apis/config/application.conf.ctmpl (hunk 1 above)
Comment (bskinner) [reply pepper-apis/config/application.conf.ctmpl]: Even if it's static, why is the URL using the IP directly and not sticking with the domain name?

File: pepper-apis/src/appengine/DataDonationPlatform.yaml [hunk 3]
```diff
@@ -13,8 +13,10 @@ automatic_scaling:
   max_concurrent_requests: 10 #default is 10 . Max is 80. GCP recommends 10 max value.
   max_pending_latency: 60ms #Default is 30ms. Max time a request is allowed to wait in queue.
   min_pending_latency: 60ms #Default is 30ms. Threshold used to scale down
-network:
-  instance_tag: study-server # necessary for firewall rules so that DSM will accept https requests from pepper
+
+vpc_access_connector:
+  # temporary proof-of-concept.  how to parameterize $PROJECT_ID without copy-pasting yaml files?
+  name: "projects/broad-ddp-dev/locations/us-central1/connectors/appengine-default-connect"
```
Comment (bskinner) [inline pepper-apis/src/appengine/DataDonationPlatform.yaml:19]: If I'm understanding your intent correctly- does the app engine configuration support substituting values from environment variables?

File: pepper-apis/src/appengine/DataDonationPlatform.yaml (hunk 2 above)
Comment (zyme) [reply pepper-apis/src/appengine/DataDonationPlatform.yaml]: This needs some help...

File: pepper-apis/src/appengine/DataDonationPlatform.yaml (hunk 3 above)
Comment (zyme) [reply pepper-apis/src/appengine/DataDonationPlatform.yaml:19]: I don't think so...might need a file per env?  CC @MocanaAtBroad

File: pepper-apis/src/appengine/DataDonationPlatform.yaml (hunk 2 above)
Comment (MocanaAtBroad) [reply pepper-apis/src/appengine/DataDonationPlatform.yaml]: YAML does  support references (yeah!) but you can't do concatenation with the referred values (boo!). I kept looking for workarounds but there aren't any with YAML alone.
This smells like a file that we will need to generate from a template.

**Comment by MocanaAtBroad** [comment]
Did you consider using GAE's firewall support? https://cloud.google.com/appengine/docs/standard/python/creating-firewalls

At least at first glance it appears that GAE's firewalls are targeting this very specific scenario.

File: pepper-apis/src/appengine/DataDonationPlatform.yaml (hunk 3 above)
Comment (MocanaAtBroad) [reply pepper-apis/src/appengine/DataDonationPlatform.yaml:19]: I believe you can set environment values in the .yaml file that become visible to the executable you are deploying. As I mentioned above, this file is a good candidate to be generated from a template.

**Comment by zyme** [comment]
@MocanaAtBroad yes, I did consider GAE firewall, but that appears to be for traffic going the other way--for inbound traffic to GAE.  In our situation, DSM will continue to run behind a separate firewall, so the need is to poke a hole in DSM's firewall to allow traffic from study-server on GAE.

**Comment by Trimax** [comment]
Discussed with @zyme. Not going to be merged.

**Pull request closed without merging**